msaad_entraid.go

```diff
@@ -0,0 +1,444 @@
+	//var errGlobal error
+	selectURL := "https://graph.microsoft.com/v1.0/users/" + user.profile.ID + "/appRoleAssignments"
+	for selectURL != "" {
+		if errGlobal != nil {
+			mp.log.Errorf("(%d) Global error detected in threadGroup-user-next loop...\n", i)
+			break
+		}
+		if mp.IsShuttingDown() {
+			break
+		}
+		j := msaadRolesJSON{}
+		err := mp.fetchJSON(selectURL, &j)
+		if err != nil {
+			errGlobal = err
+			return errGlobal, true
+		}
+		if mp.parent.Config().Verbose {
+			mp.log.Infof("User %v (%v): %v\n", user.profile.bestEmail(), user.profile.ID, j)
+			for _, u := range j.Value {
+				mp.log.Infof("%v MSAAD User Permission: (%v)", user.profile.bestEmail(), u)
+			}
+		}
+		user.roles = append(user.roles, j.Value...)
+		selectURL = j.NextLink
+	}
+	return errGlobal, false
+}
+
+func (mp *MSAADProvider) GetAADUsers() ([]*msaadUser, error) {
+	selectURL := "https://graph.microsoft.com/v1.0/users?$select=id,displayName,givenName,surname,mobilePhone,userPrincipalName,mail"
+	aadUsers := []*msaadUser{}
+	for selectURL != "" {
+		if mp.parent.Config().Verbose {
+			mp.log.Infof("Fetching %v\n", selectURL)
+		}
+		j := msaadUsersJSON{}
+		if err := mp.fetchJSON(selectURL, &j); err != nil {
+			return nil, err
+		}
+		for _, v := range j.Value {
+			aadUsers = append(aadUsers, &msaadUser{
+				profile: *v,
+			})
+		}
+		selectURL = j.NextLink
+	}
+	return aadUsers, nil
+}
+
+func (mp *MSAADProvider) fetchJSON(fetchURL string, jsonRoot interface{}) error {
+	request, err := http.NewRequest("GET", fetchURL, nil)
+	if err != nil {
+		return fmt.Errorf("Error creating Request object for url '%v': %v", fetchURL, err)
+	}
+	response, body, err := mp.doLoggedHTTP(request)
+	if err != nil {
+		return fmt.Errorf("Error fetching '%v' (err): %w", fetchURL, err)
+	}
+	if response.StatusCode != 200 {
+		return fmt.Errorf("Error fetching '%v' (response): %v", fetchURL, response.Status)
+	}
+	// if strings.Index(fetchURL, "appRole") != -1 && len(body) > 180 {
+	// 	fmt.Printf("Body: %v\n", string(body))
+	// }
+	return json.Unmarshal(body, jsonRoot)
+}
+
+// Execute doHTTP, and log any failure
+// In addition, this function reads the response body, and returns
+func (mp *MSAADProvider) doLoggedHTTP(request *http.Request) (*http.Response, []byte, error) {
+	response, err := mp.doHTTP(request)
+	var responseBody []byte
+	if err == nil && response.Body != nil {
+		defer response.Body.Close()
+		responseBody, err = io.ReadAll(response.Body)
+		if err != nil {
+			mp.log.Errorf("MSAAD failed to read response body from %v %v: %v", request.Method, request.URL.String(), err)
+			return response, nil, err
+		}
+	}
+	if err != nil {
+		mp.log.Errorf("MSAAD failed to %v %v (err): %v", request.Method, request.URL.String(), err)
+	} else if response.StatusCode != 200 {
+		mp.log.Errorf("MSAAD failed to %v %v (response): %v %v", request.Method, request.URL.String(), response.Status, string(responseBody))
+	}
+	return response, responseBody, err
```

Comment:
I don't know why this method is so hard to read... please just change it to this.
```suggestion
	response, err := mp.doHTTP(request)
	if err != nil {
		e := fmt.Errorf("MSAAD failed to %v %v (err): %w", request.Method, request.URL.String(), err)
		mp.log.Error(e.Error())
		return nil, nil, e
	}

	defer response.Body.Close()
	body, err := io.ReadAll(response.Body)
	if err != nil {
		e := fmt.Errorf("MSAAD failed to read response body from %v %v: %w", request.Method, request.URL.String(), err)
		mp.log.Error(e.Error())
		return nil, nil, e
	}

	if response.StatusCode != 200 {
		e := fmt.Errorf("MSAAD failed to %v %v (response): %v %v", request.Method, request.URL.String(), response.Status, string(responseBody))
		mp.log.Error(e.Error())
		return nil, nil, e
	}

	return response, body, nil
```
This also has the benefit of returning much better errors.